Add support for Up (Australian Bank)

ZeroEcks opened the pull request:
Hi, the [Australian bank Up](up.com.au) provides [an API](https://developer.up.com.au/), and your lovely project structure made it a breeze for me to add support.

This works for me, and I plan on running this myself, would be great to have it upstreamed so I don't have to maintain a fork of the whole repository.

Thanks!

# TODO

- [x] Remove useless type definitions (I got a little crazy with it)
- [x] Add some way to import Ups fairly intelligent categorisation with a flag?

tim-smart commented on src/Bank/Up.ts:
```diff
@@ -0,0 +1,160 @@
+
+const URL = "https://api.up.com.au/api/v1"
+
+export const UpBank = Effect.gen(function* () {
```
```suggestion
export const UpLive = Effect.gen(function* () {
```

Just to be consistent with the Bnz bank.

tim-smart commented on src/Bank/Up.ts:
```diff
@@ -0,0 +1,160 @@
+class Transaction extends S.Class<Transaction>("Transaction")({
+  type: S.Literal("transactions"),
+  id: S.String,
+  attributes: S.Struct({
```
I would remove any fields you aren't planning on using, as it means if the bank changes the format slightly the schema won't break :)

tim-smart reviewed (approved):
Nice! Almost looks good to go.

tim-smart commented:
I have added a "category" field to bank transactions, and if a matching category is found in actual it will add it to the imported transaction.

I have not tested it, so give it a go and see if there is any issues.

ZeroEcks commented:
I messed up a bit on rewriting the git history which led to the interesting history... let me know and I can fix it but I just rebased slightly further back when rewriting my username than I should have.

Anyway, I quickly added some logic to match the transactions when transferring between accounts inside the same bank, you just have to set the "transfer" field. I also added mapping from the bank categories (`public-transport`) to actual categories (`transit`) because the bank categories aren't really human readable, and are far more specific than you would want.

How do we feel about this structure? I refactored `Sync.ts` a little to modularise the optional properties (transfers and category) because I could feel it getting messy.

IMO we should add some tests, because I am not sure if my changes could potentially break the NZ bank, and you wouldn't be sure if you broke Up. I have no further opinions on the matter, but I would implement some tests for it if we had a testing framework setup.

I think a nice feature to have in the future would be a wizard to generate the command args by listing the accounts and asking you to pair them up, along with a simple interface to create mappings for the categories. For me the final destination of this command would just be in a kubernetes cron job, so I only really need to generate the args once. Otherwise it might be worth considering what the actual design of passing the configuration would be? Maybe it should instead read a configuration file?

tim-smart commented:
> I think a nice feature to have in the future would be a wizard to generate the command args by listing the accounts and asking you to pair them up, along with a simple interface to create mappings for the categories. For me the final destination of this command would just be in a kubernetes cron job, so I only really need to generate the args once. Otherwise it might be worth considering what the actual design of passing the configuration would be? Maybe it should instead read a configuration file?

Yes could be a nice feature. Currently you can run --wizard to generate a command, but it isn't very helpful for key-value maps right now.
